tango/integrations/beaker/step_cache.py

```diff
@@ -0,0 +1,178 @@
+            return dataset if dataset.committed is not None else None
+        except DatasetNotFound:
+            return None
+
+    def _sync_step_dataset(self, step: Step, objects_dir: Path) -> Dataset:
+        dataset_name = step_dataset_name(step)
+        try:
+            dataset = self.beaker.dataset.create(dataset_name, commit=False)
+        except DatasetConflict:
+            dataset = self.beaker.dataset.get(dataset_name)
+
+        try:
+            self.beaker.dataset.sync(dataset, objects_dir, quiet=True)
+            dataset = self.beaker.dataset.commit(dataset)
+        except DatasetWriteError:
+            pass
+
+        return dataset
+
+    def __contains__(self, step: Any) -> bool:
+        if isinstance(step, (Step, StepInfo)):
+            cacheable = step.cache_results if isinstance(step, Step) else step.cacheable
+            if not cacheable:
+                return False
+
+            key = step.unique_id
+
+            # First check if we have a copy in memory.
+            if key in self.strong_cache:
+                return True
+            if key in self.weak_cache:
+                return True
+
+            # Then check if we have a copy on disk in our cache directory.
+            with self._acquire_step_lock_file(step, read_only_ok=True):
+                if (self.step_dir(step) / Constants.STEP_RESULT_DIR).is_dir():
+                    return True
+
+            # If not, check Beaker for the corresponding dataset.
+            return self._step_result_dataset(step) is not None
+        else:
+            return False
+
+    def __getitem__(self, step: Union[Step, StepInfo]) -> Any:
+        key = step.unique_id
+
+        # Try getting the result from our in-memory caches first.
+        result = self._get_from_cache(key)
+        if result is not None:
+            return result
+
+        def load_and_return():
+            metadata = CacheMetadata.from_params(Params.from_file(self._metadata_path(step)))
+            result = metadata.format.read(self.step_dir(step) / Constants.STEP_RESULT_DIR)
+            self._add_to_cache(key, result)
+            return result
+
+        # Next check our local on-disk cache.
+        with self._acquire_step_lock_file(step, read_only_ok=True):
+            if self.step_dir(step).is_dir():
+                return load_and_return()
+
+        # Finally, check Beaker for the corresponding dataset.
+        with self._acquire_step_lock_file(step):
+            # Make sure the step wasn't cached since the last time we checked (above).
+            if self.step_dir(step).is_dir():
+                return load_and_return()
+
+            dataset = self._step_result_dataset(step)
+            if dataset is None:
+                raise KeyError(step)
+
+            # We'll download the dataset to a temporary directory first, in case something goes wrong.
+            temp_dir = tempfile.mkdtemp(dir=self.dir, prefix=key)
```

Comment:
This is fine, but `tempfile.TemporaryDirectory` exists and gives a higher-level API.

Reply:
Ah, I guess `TemporaryDirectory` doesn't have `rmtree` capability.